File: Competitive-Programming/LeetCode/golang/main/P121.go
package main


import (
	"fmt"
)

func maxProfit(prices []int) int {
	if len(prices) <= 1 {
		return 0
	}

	min := prices[0]
	maxProfit := 0

	for i := 1; i < len(prices); i++ {
		thisPrice := prices[i]
		if thisPrice < min {
			min = thisPrice
		}

		thisProfit := thisPrice - min
		if thisProfit > maxProfit {
			maxProfit = thisProfit
		}
	}

	return maxProfit
}

func check(b bool) {
	if !b {
		fmt.Println("error")
	}
}

func main() {
	check(maxProfit([]int{1, 2}) == 1)
	check(maxProfit([]int{3, 3, 5, 0, 0, 3, 1, 4}) == 4)
	check(maxProfit([]int{7, 1, 5, 3, 6, 4}) == 5)
}

File: Competitive-Programming/LeetCode/python/P121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        min = 100000
        for n in prices:
            if n < min:
                min = n
            elif n - min > profit:
                profit = n - min
        return profit

            

s = Solution()

prices = [7,1,5,3,6,4]
a = s.maxProfit(prices)

prices = [5,4,3,2]
a = s.maxProfit(prices)
        